coconut/src/request.rs

```diff
@@ -0,0 +1,367 @@
+use crate::error::{CoconutError, CoconutResult};
+use crate::issue::Coin;
+use crate::setup::{Parameters, PublicKey};
+use bls12_381::{G1Projective, G2Projective, Scalar};
+use group::GroupEncoding as _;
+use sha2::{Digest as _, Sha512};
+use std::convert::TryInto;
+
+#[cfg(test)]
+#[path = "tests/request_tests.rs"]
+pub mod request_tests;
+
+pub struct CoinsRequest {
+    /// Input credentials representing coins.
+    sigmas: Vec<Coin>,
+    /// Kappa group elements associated with the input credentials (`sigmas`).
+    kappas: Vec<G2Projective>,
+    /// Nu group elements associated with the input credentials (`sigmas`).
+    nus: Vec<G1Projective>,
+    /// The common commitments Cm of the output coin values and ids.
+    pub cms: Vec<G1Projective>,
+    /// The blinded output coin values and ids.
+    pub cs: Vec<(G1Projective, G1Projective)>,
+    /// A ZK-proof asserting correctness of all the other fields and that the sum of the input
+    /// coins equals the sum of the output coins.
+    proof: RequestCoinsProof,
+}
+
+impl CoinsRequest {
+    pub fn new(
+        // The system parameters.
+        parameters: &mut Parameters,
+        // The aggregated public key of the authorities.
+        public_key: &PublicKey,
+        // The credentials representing the input coins. Each credential has two attributes, a coin
+        // value and a id.
+        sigmas: &[Coin],
+        // The attributes of the credentials `sigmas`. Each input attribute is a tuple of (coin value, id).
+        input_attributes: &[(Scalar, Scalar)],
+        // Each output attribute is a tuple of (coin value, id).
+        output_attributes: &[(Scalar, Scalar)],
+        // Each element contains one blinding factor for the coin value and one for the id. There should
+        // as many blinding factors as output attributes.
+        blinding_factors: &[(Scalar, Scalar)],
+    ) -> Self {
+        debug_assert!(sigmas.len() == input_attributes.len());
```

Comment:
So far, I've used just `assert!` (or `ensure!`) in this codebase. I kinda like the general note about this in the Rust library:
https://doc.rust-lang.org/std/macro.debug_assert.html#:~:text=Unlike%20assert!%20%2C%20debug_assert!,non%20optimized%20builds%20by%20default.&text=The%20performance%20cost%20of%20assertions,importantly%2C%20only%20in%20safe%20code!

Reply:
Yes, but a general advice by the Rust community is to prefer `assert!` over `debug_assert!` (unless you have a good reason) so that you test the production behavior.